### PR TITLE
Collecting dump during SAI failure

### PR DESCRIPTION
**Pull request by dgsudharsan**
This HLD is to describe the flow to collect useful dumps during SAI failures.
| Repo  | PR title | State |
| ---------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- |
| sonic-swss | [[sai_failure_dump]Invoking dump during SAI failure](https://github.com/sonic-net/sonic-swss/pull/2644) | GitHub issue/pull request detail |
| sonic-sairedis | [[sai_failure_dump]Invoking dump during SAI failure](https://github.com/sonic-net/sonic-sairedis/pull/1198)  | GitHub issue/pull request detail |  
| sonic-utilities | [[sai_failure_dump]Invoking dump during SAI failure](https://github.com/sonic-net/sonic-utilities/pull/2633 ) | GitHub issue/pull request detail |
| sonic-buildimage   | [[Mellanox][sai_failure_dump]Added platform specific script to be invoked during SAI failure dump](https://github.com/sonic-net/sonic-buildimage/pull/13533) | GitHub issue/pull request detail |

### CODE REVIEW

File: doc/SAI_failure_handling/dump_on_sai_failure.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+### Scope
+The scope of this document is to design the handling of taking a dump during a SAI failure. 
+
+### Definitions/Abbreviations
+ 
+
+### Overview
+In the existing design, when SAI failure occurs, the orchagent aborts and all the dependent services including syncd restart. This results in failure to take the SAI, SDK and lower layer state during the problem state resulting in loss of information to debug.
+To solve this issue, whenever there is a SAI failure, orchagent requests syncd to take relevant dumps and once done, it proceeds for the abort.
```
Comment (prvattem) [inline doc/SAI_failure_handling/dump_on_sai_failure.md:37]: Are the contents of dumps at various layers, implementation/platform specific or is there a plan to formalize these ?

File: doc/SAI_failure_handling/dump_on_sai_failure.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+### Requirements
+
+Primary requirements for taking dump during SAI failure are
+- The dump needs to be taken synchronosly before abort.
```
Comment (prvattem) [inline doc/SAI_failure_handling/dump_on_sai_failure.md:42]: how much delay is acceptable before calling abort?

File: doc/SAI_failure_handling/dump_on_sai_failure.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+
+### High-Level Design
+A new custom attribute SAI_REDIS_SWITCH_ATTR_FAILURE_DUMP is defined. When there is a SAI failure, before calling the abort, orchagent sets the SAI_REDIS_SWITCH_ATTR_FAILURE_DUMP attribute. On receiving this attribute syncd calls the generic dump script which is present in /usr/bin/syncd_dump.sh. This script will check for the presence of platform specific dump script which should be located at /usr/bin/platform_syncd_dump.sh. If this script is present, it would be invoked to take the necessary dump. Vendors if they intend to take dumps during SAI failure can define the script in their syncd docker. The dumps collected from this script should be stored in /var/log/sai_failure_dump/ which will be exposed to the host. Only one file should be stored per dump in order to facilitate the rotation logic. Once the dump is finished, the generic syncd dump script will perform rotation on /var/log/sai_failure_dump/ to restrict the number of dumps. A variable by name SAI_MAX_FAILURE_DUMPS is defined in the generic script which by default is set to 10. This variable can be overwritten in the platform specific script if the platform wants a different number of dumps.
```
Comment (prvattem) [inline doc/SAI_failure_handling/dump_on_sai_failure.md:49]: When is SAI_REDIS_SWITCH_ATTR_FAILURE_DUMP cleared?

File: doc/SAI_failure_handling/dump_on_sai_failure.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+Later when techsupport is invoked manually or invoked through auto techsupport, these dumps will be collected and once collected, they will be cleared from /var/log/sai_failure_dump/
+
+The above path will also be part of logrotate script logic inorder to have a check on the dumps not over running the disk space.
+
+The below diagram explains the sequence when a SAI failure happens
+![](/images/SAI_failure_handling/SAI_failure_dump_sequence.JPG)
+
+The flow inside syncd is shown below
+![](/images/SAI_failure_handling/SAI_failure_dump_flow.JPG)
```
Comment (prvattem) [inline doc/SAI_failure_handling/dump_on_sai_failure.md:59]: Please tag all the flows

**Comment by vishnushetty** [comment]
There is no clean API interface to get the debug / tech support dump from SAI. Whatever proposed in this HLD  is specific to device/vendor specific how they collect dump using the platform script. Instead it would be cleaner define SAI interface (may be SWITCH attribute level) to collect the debug dump. OA/syncd will retrieve the dump using SAI interface whenever there is feature failure.  As per today discussion this needs to be taken up in SAI forum.

File: doc/SAI_failure_handling/dump_on_sai_failure.md (hunk 2 above)
Comment (dgsudharsan) [reply doc/SAI_failure_handling/dump_on_sai_failure.md:42]: There is no requirement over there. By default it is the timeout of syncd orchagent channel timeout (which is 1 minute)

File: doc/SAI_failure_handling/dump_on_sai_failure.md (hunk 1 above)
Comment (dgsudharsan) [reply doc/SAI_failure_handling/dump_on_sai_failure.md:37]: That needs to be a different feature. For now we provide flexibility for the vendors to include their scripts. If this needs to be formalized, it needs to be take in the SAI forum

File: doc/SAI_failure_handling/dump_on_sai_failure.md (hunk 3 above)
Comment (dgsudharsan) [reply doc/SAI_failure_handling/dump_on_sai_failure.md:49]: A minor correction. I updated the attribute to use existing SAI_REDIS_SWITCH_ATTR_NOTIFY_SYNCD with value SAI_REDIS_NOTIFY_SYNCD_INVOKE_DUMP The notification attribute is immediately cleared after syncd handles them

File: doc/SAI_failure_handling/dump_on_sai_failure.md (hunk 4 above)
Comment (dgsudharsan) [reply doc/SAI_failure_handling/dump_on_sai_failure.md:59]: Can you please elaborate what do you mean by tag?

Comment (prvattem) [reply doc/SAI_failure_handling/dump_on_sai_failure.md:59]: Flow chart flow needs to be labelled for better understanding.
For example "if platform dump script exists"   --> No --> Perform Dump Rotate
                                                                          --> Yes --> Invoke Platform specific script --> Save dump.

Comment (dgsudharsan) [reply doc/SAI_failure_handling/dump_on_sai_failure.md:59]: Thanks. Updated

File: doc/SAI_failure_handling/dump_on_sai_failure.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (dgsudharsan) [reply doc/SAI_failure_handling/dump_on_sai_failure.md]: I tested with 12K routes which is the maximum scale in sonic-mgmt tests and it took 5 seconds. Apparently for our platforms, our saisdkdump is not affected by scale as it is a memory dump.

**Comment by liat-grozovik** [comment]
@prsunny  kindly reminder to comment/approve or assign someone on your behalf

**Review by prsunny — APPROVED** [review]

**Review by prvattem — APPROVED** [review]

**Pull request merged**